File: subarray-sum-equals-k/subarray-sum-equals-k.py
"""
time: O(n)
space: O(n) for auxiliary dict
"""
from typing import Dict

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        # using a map to store the sum as keys and the number of occurrences of that sum as v
        d: Dict[int,int] = {}
        d[0] = 1 # there is a value that sums to 0; e.g. the absence of a val
        cum_sum = 0
        total = 0
        for v in nums:
            cum_sum += v 
            complement = cum_sum - k
            if complement in d:
                total += d[complement]
            d[cum_sum] = d.get(cum_sum, 0) + 1
            
        return total
        

File: subarray-sum-equals-k/subarray-sum-equals-k.go
//prefix sum with hashmap approach
//constraint: 1 <= nums.length <= 2*10^4
//O(n) time and space
func subarraySum(nums []int, k int) int {
    //cnt continuous subarrays with sum = k
    cnt := 0
    runningSum := 0
    sumMap := make(map[int]int)
    sumMap[0] = 1 //cnt of sum 0 is 1
    for i:=0; i < len(nums); i++ {
        v := nums[i]
        runningSum += v
        if _, ok :=  sumMap[runningSum-k]; ok {
            cnt += sumMap[runningSum-k]
        }
        sumMap[runningSum]++
    }
    return cnt
}